### PR TITLE
Generation of Dates, Part 2

### PR DESCRIPTION
**Pull request by zinki97**
See https://github.com/jlink/jqwik/issues/140

### CODE REVIEW

**Comment by zinki97** [comment]
"All checks have failed"

What is the problem?
```
Run scacap/action-surefire-report@v1
Going to parse results form **/build/test-results/test/TEST-*.xml
Result: 1435 tests run, 0 skipped, 0 failed.
Posting status 'completed' with conclusion 'success' to https://github.com/jlink/jqwik/pull/147 (sha: 748d403a4230ebf3582e4076f27098bca5fa5eda)
Error: Resource not accessible by integration
```

File: time/src/test/java/net/jqwik/time/api/YearTests.java [hunk 1]
```diff
@@ -45,6 +46,27 @@ void yearIsNotZero(@ForAll("yearsAround0") Year year) {
 
 	}
 
+	@Group
+	class validYearsAreGeneratedWithAnnotation {
```
Comment (jlink) [inline time/src/test/java/net/jqwik/time/api/YearTests.java:50]: jqwik convention is to have Group classes be upper case. Use `@Label("myname")` if you want the displayed test name to be different.

**Comment by jlink** [comment]
The problem is that your Github user does (whose pull request invoked the
GH action to run) not have the rights to write a test report. I might
research how I can fix that, but not today.

Am Mo., 11. Jan. 2021 um 16:04 Uhr schrieb zinki97 <notifications@github.com

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultPeriodArbitrary.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	@Override
+	protected Arbitrary<Period> arbitrary() {
+
+		IntegerArbitrary years = Arbitraries.integers().between(yearsMin, yearsMax);
+		IntegerArbitrary months = Arbitraries.integers().between(monthsMin, monthsMax);
+		IntegerArbitrary days = Arbitraries.integers().between(daysMin, daysMax);
+
+		return Combinators.combine(years, months, days).as((y, m, d) -> Period.ofYears(y).plusMonths(m).plusDays(d));
+
+	}
```
Comment (zinki97) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultPeriodArbitrary.java:34]: @jlink Can you please provide feedback to this type of generation of periods?

- Maybe days should be between 0 and 30 by default?
- Maybe months should be between 0 and 11 by default?

Comment (jlink) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultPeriodArbitrary.java:34]: I think those defaults are reasonable

File: time/src/test/java/net/jqwik/time/api/PeriodTests.java [hunk 3]
```diff
@@ -126,12 +144,124 @@ void daysBetweenSame(@ForAll int start, @ForAll Random random) {
...
+		}
+
+		@Disabled
+		@Property
+		void shrinksToSmallestFailingValue(@ForAll Random random) {
+			PeriodArbitrary periods = Dates.periods();
+			TestingFalsifier<Period> falsifier = period -> comparePeriodsWithMax30DaysAnd11Months(period, Period.of(200, 3, 5)) <= 0;
+			Period value = falsifyThenShrink(periods, random, falsifier);
+			assertThat(value).isEqualTo(Period.of(200, 3, 5));
+		}
```
Comment (zinki97) [inline time/src/test/java/net/jqwik/time/api/PeriodTests.java:164]: Again it doesn't shrink perfectly. It shrinks to `Period.of(201, 0, 0)`

Comment (jlink) [reply time/src/test/java/net/jqwik/time/api/PeriodTests.java:164]: Using only days (and calculating it from days, months and days) would improve that.

Comment (zinki97) [reply time/src/test/java/net/jqwik/time/api/PeriodTests.java:164]: Yeah, but the problem here is, that you can't calculate it from days I guess:

- years/months/days can be negative and positive (if the user set it), so there are more solutions for one value
- months have different length and you don't know which month it is

Comment (zinki97) [reply time/src/test/java/net/jqwik/time/api/PeriodTests.java:164]: I have an idea how you can calculate it from one number, but this number has to be very large. I'll try it

Comment (jlink) [reply time/src/test/java/net/jqwik/time/api/PeriodTests.java:164]: If it doesn't work we have to think of something else.

Comment (jlink) [reply time/src/test/java/net/jqwik/time/api/PeriodTests.java:164]: After experimenting a bit with the Period API and how it's supposed to work, I guess your first approach was right and I'll have to tweak shrinking a bit.

File: time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultPeriodArbitrary.java [hunk 4]
```diff
@@ -15,21 +16,27 @@
...
-		IntegerArbitrary years = Arbitraries.integers().between(yearsMin, yearsMax);
-		IntegerArbitrary months = Arbitraries.integers().between(monthsMin, monthsMax);
-		IntegerArbitrary days = Arbitraries.integers().between(daysMin, daysMax);
+		BigInteger bigIntegerStart = BigInteger.ZERO;
+		BigInteger bigIntegerEnd = calculateBigIntegerEnd();
+
+		BigIntegerArbitrary numbers = Arbitraries.bigIntegers().between(bigIntegerStart, bigIntegerEnd);
 
-		Arbitrary<Period> periodArbitrary = Combinators.combine(years, months, days).as(Period::of);
+		Arbitrary<Period> periodArbitrary = numbers.map(this::calculatePeriod);
```
Comment (zinki97) [inline time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultPeriodArbitrary.java]: Here is another solution. Shrinking works, but equal distribution fails (positive and negative)

```
timestamp = 2021-01-22T13:01:25.106151200, [CheckEqualDistribution:periodCanBePositiveAndNegative] (1000) Period is negative = 
    true  (953) : 95 %
    false ( 47) :  5 %
```

Which solution is better?

Comment (zinki97) [reply time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultPeriodArbitrary.java]: `Arbitrary<BigInteger> numbers = Arbitraries.bigIntegers().withDistribution(RandomDistribution.uniform()).between(bigIntegerStart, bigIntegerEnd);`

Found a solution :)

**Comment by jlink** [comment]
> The problem is that your Github user does (whose pull request invoked the GH action to run) not have the rights to write a test report. I might research how I can fix that, but not today. 

Test report is no longer created for pull requests so that problem should have disappeared.

File: time/src/main/java/net/jqwik/time/api/Dates.java [hunk 5]
```diff
@@ -81,4 +100,13 @@ public static MonthDayArbitrary monthDays() {
 		return new DefaultMonthDayArbitrary();
 	}
 
+	/**
+	 * Create an arbitrary that generates instances of {@linkplain java.time.Period}.
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/api/Dates.java:104]: Maybe you can add a short note about the possible ranges of periods being generated.

**Comment by zinki97** [comment]
.* Request review *

File: documentation/src/docs/include/additional-modules.md [hunk 6]
```diff
@@ -74,69 +74,87 @@ This module is part of jqwik's default dependencies.
...
+##### PeriodArbitrary
+
+- You can constrain the minimum and maximum value for years using `yearsBetween(min, max)` and `@PeriodYearRange`.
+- You can constrain the minimum and maximum value for months using `monthsBetween(min, max)` and `@PeriodMonthRange`.
+- You can constrain the minimum and maximum value for days using `daysBetween(min, max)` and `@PeriodDayRange`.
+- By default, years between `Integer.MIN_VALUE` and `Integer.MAX_VALUE`, months between `0` and `11` and days between `0` and `30` are generated.
+
+##### Special note on `@DateRange`, `@MonthDayRange` and `@YearMonthRange`
+
+`@DateRange`, `@MonthDayRange` and `@YearMonthRange` use the ISO format for date strings. Examples: `2013-05-25`, `05-25` and `2013-05`.
```
Comment (jlink) [inline documentation/src/docs/include/additional-modules.md:156]: Shouldn't a month-day ISO string have a leading "-", e.g. "-05-21" ?

File: documentation/src/docs/include/additional-modules.md [hunk 7]
```diff
@@ -74,69 +74,87 @@ This module is part of jqwik's default dependencies.
```
Comment (jlink) [inline documentation/src/docs/include/additional-modules.md:74]: Show at least one example for generating dates. Compare with annotation to provide method.

File: time/src/main/java/net/jqwik/time/internal/properties/configurators/YearMonthRangeConfigurator.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+		}
+	}
+
+	private YearMonth isoDateToYearMonth(String iso) {
+		if (iso == null || iso.length() == 0) {
+			return null;
+		}
+		String[] parts = iso.split("-");
+		if (parts.length != 2) {
+			return null;
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/configurators/YearMonthRangeConfigurator.java:33]: Throw exception when format is wrong.

File: time/src/main/java/net/jqwik/time/internal/properties/configurators/DateRangeConfigurator.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+		return calendar.getTime();
+	}
+
+	private LocalDate isoDateToLocalDate(String iso) {
+		if (iso == null || iso.length() == 0) {
+			return null;
+		}
+		String[] parts = iso.split("-");
+		if (parts.length != 3) {
+			return null;
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/configurators/DateRangeConfigurator.java:58]: Throw exception when format is wrong.

File: time/src/main/java/net/jqwik/time/internal/properties/configurators/MonthDayRangeConfigurator.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+			MonthDayArbitrary monthDayArbitrary = (MonthDayArbitrary) arbitrary;
+			return monthDayArbitrary.between(isoDateToMonthDay(range.min()), isoDateToMonthDay(range.max()));
+		} else {
+			return arbitrary;
+		}
+	}
+
+	private MonthDay isoDateToMonthDay(String iso) {
+		if (iso == null || iso.length() == 0) {
+			return null;
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/internal/properties/configurators/MonthDayRangeConfigurator.java:29]: Throw exception when format is wrong.

File: time/src/main/java/net/jqwik/time/internal/properties/configurators/YearMonthRangeConfigurator.java (hunk 8 above)
Comment (zinki97) [reply time/src/main/java/net/jqwik/time/internal/properties/configurators/YearMonthRangeConfigurator.java:33]: How can I test that this throws an exception?

```
@Example
void dateRangeThrowsException(@ForAll @DateRange(min = "2013-05") Date date){
	assertThatExceptionOfType(DateTimeParseException.class); //not working
}
```

```
timestamp = 2021-01-25T20:05:28.758341700, DateConstraints:dateRangeThrowsException = 
  java.time.format.DateTimeParseException:
    Test
```

Comment (zinki97) [reply time/src/main/java/net/jqwik/time/internal/properties/configurators/YearMonthRangeConfigurator.java:33]: I've found a solution after experimenting a bit :)

```
@Example
@ExpectFailure(failureType = DateTimeParseException.class)
void dateRangeThrowsException(@ForAll @DateRange(min = "2013-05") Date date){
	//do nothing
}
```

**Comment by zinki97** [comment]
.* Request review *

**Pull request merged**